File: backend-rust/src/exchanges/combined.rs
use crate::state::AppState;
use serde_json::{json, Value};
use std::collections::HashMap;
use std::sync::Arc;

struct MergedContract {
    strike: f64,
    option_type: String,
    best_bid: f64,
    best_bid_ex: String,
    best_ask: f64,
    best_ask_ex: String,
    prices: HashMap<String, (f64, f64)>, // exchange -> (bid, ask)
    delta: f64,
    gamma: f64,
    theta: f64,
    vega: f64,
    mark_vol: f64,
    bid_vol: f64,
    ask_vol: f64,
}

impl MergedContract {
    fn new(strike: f64, option_type: String) -> Self {
        let mut prices = HashMap::new();
        for ex in &["bybit", "okx", "deribit", "derive", "binance"] {
            prices.insert(ex.to_string(), (0.0, 0.0));
        }
        Self {
            strike,
            option_type,
            best_bid: 0.0,
            best_bid_ex: String::new(),
            best_ask: 0.0,
            best_ask_ex: String::new(),
            prices,
            delta: 0.0,
            gamma: 0.0,
            theta: 0.0,
            vega: 0.0,
            mark_vol: 0.0,
            bid_vol: 0.0,
            ask_vol: 0.0,
        }
    }

    fn to_value(&self) -> Value {
        let prices_obj: serde_json::Map<String, Value> = self
            .prices
            .iter()
            .map(|(ex, (bid, ask))| {
                (
                    ex.clone(),
                    json!({ "bid": bid, "ask": ask }),
                )
            })
            .collect();

        json!({
            "strike":     self.strike,
            "optionType": self.option_type,
            "bestBid":    self.best_bid,
            "bestBidEx":  self.best_bid_ex,
            "bestAsk":    self.best_ask,
            "bestAskEx":  self.best_ask_ex,
            "prices":     prices_obj,
            "delta":      self.delta,
            "gamma":      self.gamma,
            "theta":      self.theta,
            "vega":       self.vega,
            "markVol":    self.mark_vol,
            "bidVol":     self.bid_vol,
            "askVol":     self.ask_vol,
        })
    }
}

fn parse_f64(v: &Value) -> f64 {
    match v {
        Value::Number(n) => n.as_f64().unwrap_or(0.0),
        Value::String(s) => s.parse().unwrap_or(0.0),
        _ => 0.0,
    }
}

fn merge_exchange_data(
    merged: &mut HashMap<String, MergedContract>,
    forward_prices: &mut HashMap<String, f64>,
    exchange_name: &str,
    response: &Value,
    bid_multiplier: f64,  // for OKX: spot_price; for others: 1.0
) {
    let data = match response["data"].as_object() {
        Some(d) => d,
        None => return,
    };

    for (expiry, expiry_data) in data {
        // forwardPrice: first non-zero wins
        let fwd = parse_f64(&expiry_data["forwardPrice"]);
        if fwd > 0.0 {
            forward_prices.entry(expiry.clone()).or_insert(fwd);
        }

        for option_type in &["calls", "puts"] {
            let contracts = match expiry_data[option_type].as_array() {
                Some(arr) => arr,
                None => continue,
            };

            for contract in contracts {
                let strike = parse_f64(&contract["strike"]);
                let opt_type = contract["optionType"].as_str().unwrap_or("").to_string();
                let key = format!("{}|{}|{}", expiry, strike, opt_type);

                let entry = merged
                    .entry(key)
                    .or_insert_with(|| MergedContract::new(strike, opt_type));

                let raw_bid = parse_f64(&contract["bid"]);
                let raw_ask = parse_f64(&contract["ask"]);
                let bid = raw_bid * bid_multiplier;
                let ask = raw_ask * bid_multiplier;

                // Update per-exchange prices
                entry.prices.insert(exchange_name.to_string(), (bid, ask));

                // bestBid: highest bid wins
                if bid > entry.best_bid {
                    entry.best_bid = bid;
                    entry.best_bid_ex = exchange_name.to_string();
                }

                // bestAsk: lowest ask wins (ask > 0)
                if ask > 0.0 && (entry.best_ask == 0.0 || ask < entry.best_ask) {
                    entry.best_ask = ask;
                    entry.best_ask_ex = exchange_name.to_string();
                }

                // Greeks: last non-zero wins
                let delta = parse_f64(&contract["delta"]);
                let gamma = parse_f64(&contract["gamma"]);
                let theta = parse_f64(&contract["theta"]);
                let vega = parse_f64(&contract["vega"]);
                let mark_vol = parse_f64(&contract["markVol"]);
                let bid_vol = parse_f64(&contract["bidVol"]);
                let ask_vol = parse_f64(&contract["askVol"]);

                if delta != 0.0 { entry.delta = delta; }
                if gamma != 0.0 { entry.gamma = gamma; }
                if theta != 0.0 { entry.theta = theta; }
                if vega != 0.0 { entry.vega = vega; }
                if mark_vol != 0.0 { entry.mark_vol = mark_vol; }
                if bid_vol != 0.0 { entry.bid_vol = bid_vol; }
                if ask_vol != 0.0 { entry.ask_vol = ask_vol; }
            }
        }
    }
}

pub async fn build_combined_response(state: &Arc<AppState>, coin: &str) -> Option<Value> {
    let coin_upper = coin.to_uppercase();

    // Acquire all read locks and call each exchange's build_response
    let bybit_response = {
        let ticker = state.bybit_ticker.read().await;
        let spot = state.bybit_spot.read().await;
        crate::exchanges::bybit::build_response(&ticker, &spot, &coin_upper)
    };

    let okx_response = if coin_upper == "BTC" || coin_upper == "ETH" {
        let greeks = state.okx_greeks.read().await;
        let ticker = state.okx_ticker.read().await;
        let spot = state.okx_spot.read().await;
        let family = format!("{}-USD", coin_upper);
        let resp = crate::exchanges::okx::build_response(&greeks, &ticker, &spot, &family);
        // Capture spot price for bid/ask multiplication
        let spot_price = if coin_upper == "BTC" {
            spot.get("BTC-USDT").copied().unwrap_or(0.0)
        } else {
            spot.get("ETH-USDT").copied().unwrap_or(0.0)
        };
        (resp, spot_price)
    } else {
        (Value::Null, 0.0)
    };

    let deribit_response = {
        let deribit = state.deribit.read().await;
        let greeks = state.deribit_greeks.read().await;
        crate::exchanges::deribit::build_response(&deribit, &greeks, &coin_upper)
            .unwrap_or(Value::Null)
    };

    let derive_response = if coin_upper == "BTC" || coin_upper == "ETH" {
        let tickers = state.derive_tickers.read().await;
        let spot = state.derive_spot.read().await;
        crate::exchanges::derive::build_response(&tickers, &spot, &coin_upper)
            .unwrap_or(Value::Null)
    } else {
        Value::Null
    };

    let binance_response = {
        let binance = state.binance.read().await;
        let spot = state.binance_spot.read().await;
        crate::exchanges::binance::build_response(&binance, &spot, &coin_upper)
            .unwrap_or(Value::Null)
    };

    // Determine best spot price (max non-zero across exchanges)
    let mut best_spot = 0.0f64;
    for resp in &[&bybit_response, &okx_response.0, &deribit_response, &derive_response, &binance_response] {
        let s = resp["spotPrice"].as_f64().unwrap_or(0.0);
        if s > best_spot {
            best_spot = s;
        }
    }

    // Merge all exchanges
    let mut merged: HashMap<String, MergedContract> = HashMap::new();
    let mut forward_prices: HashMap<String, f64> = HashMap::new();

    // Bybit (bid/ask already in USD)
    if !bybit_response.is_null() {
        merge_exchange_data(&mut merged, &mut forward_prices, "bybit", &bybit_response, 1.0);
    }

    // OKX: bid/ask in BTC units, multiply by spot
    let (okx_resp, okx_spot) = okx_response;
    if !okx_resp.is_null() {
        let multiplier = if okx_spot > 0.0 { okx_spot } else { best_spot };
        merge_exchange_data(&mut merged, &mut forward_prices, "okx", &okx_resp, multiplier);
    }

    // Deribit (bid/ask already in USD)
    if !deribit_response.is_null() {
        merge_exchange_data(&mut merged, &mut forward_prices, "deribit", &deribit_response, 1.0);
    }

    // Derive (bid/ask already in USD)
    if !derive_response.is_null() {
        merge_exchange_data(&mut merged, &mut forward_prices, "derive", &derive_response, 1.0);
    }

    // Binance (bid/ask already in USD)
    if !binance_response.is_null() {
        merge_exchange_data(&mut merged, &mut forward_prices, "binance", &binance_response, 1.0);
    }

    if merged.is_empty() {
        return None;
    }

    // Group by expiry
    let mut by_expiry: HashMap<String, (Vec<Value>, Vec<Value>)> = HashMap::new();

    for (key, contract) in &merged {
        let expiry = key.split('|').next().unwrap_or("").to_string();
        let entry = by_expiry.entry(expiry).or_insert_with(|| (vec![], vec![]));
        let val = contract.to_value();
        if contract.option_type == "call" {
            entry.0.push(val);
        } else {
            entry.1.push(val);
        }
    }

    // Sort each expiry's contracts by strike
    for (calls, puts) in by_expiry.values_mut() {
        calls.sort_by(|a, b| {
            let sa = a["strike"].as_f64().unwrap_or(0.0);
            let sb = b["strike"].as_f64().unwrap_or(0.0);
            sa.partial_cmp(&sb).unwrap_or(std::cmp::Ordering::Equal)
        });
        puts.sort_by(|a, b| {
            let sa = a["strike"].as_f64().unwrap_or(0.0);
            let sb = b["strike"].as_f64().unwrap_or(0.0);
            sa.partial_cmp(&sb).unwrap_or(std::cmp::Ordering::Equal)
        });
    }

    let mut sorted_dates: Vec<String> = by_expiry.keys().cloned().collect();
    sorted_dates.sort();

    let mut expiration_counts = serde_json::Map::new();
    let mut data_obj = serde_json::Map::new();

    for date in &sorted_dates {
        let (calls, puts) = &by_expiry[date];
        let fwd = forward_prices.get(date).copied().unwrap_or(0.0);
        expiration_counts.insert(
            date.clone(),
            json!({ "calls": calls.len(), "puts": puts.len() }),
        );
        data_obj.insert(
            date.clone(),
            json!({
                "calls": calls,
                "puts": puts,
                "forwardPrice": fwd,
            }),
        );
    }

    Some(json!({
        "spotPrice":        best_spot,
        "expirations":      sorted_dates,
        "expirationCounts": expiration_counts,
        "data":             data_obj,
    }))
}

File: archive/backend-node/lib/combined.js
export function buildCombinedResponse(baseCoin, deps) {
  const {
    bybitApi,
    bybitTickerCache,
    bybitSpotCache,
    okxCache,
    okxTickerCache,
    okxSpotCache,
    parseOkxInstId,
    buildDeribitResponse,
    buildDeriveResponse,
    binanceCache,
    binanceSpotCache,
    parseBinanceSymbol,
  } = deps

  const okxFamily = `${baseCoin}-USD`
  const bybitTickers = Object.values(bybitTickerCache[baseCoin] ?? {})
  const bybitSpot = bybitSpotCache[baseCoin] ?? 0
  const okxFamilyCache = okxCache[okxFamily] ?? {}
  const okxFamilyTickers = okxTickerCache[okxFamily] ?? {}
  const okxSpot = okxSpotCache[`${baseCoin}-USDT`] ?? 0
  const deriveResp = buildDeriveResponse(baseCoin)
  const deriveSpot = deriveResp?.spotPrice ?? 0
  const spotPrice = bybitSpot || okxSpot || deriveSpot

  const merged = {}
  const forwardPrices = {}

  const ensure = (key) => {
    if (!merged[key]) {
      merged[key] = {
        bestBid: 0,
        bestBidEx: null,
        bestAsk: 0,
        bestAskEx: null,
        prices: {
          bybit: { bid: 0, ask: 0 },
          okx: { bid: 0, ask: 0 },
          deribit: { bid: 0, ask: 0 },
          derive: { bid: 0, ask: 0 },
          binance: { bid: 0, ask: 0 },
        },
        delta: 0,
        gamma: 0,
        theta: 0,
        vega: 0,
        markVol: 0,
        bidVol: 0,
        askVol: 0,
      }
    }
    return merged[key]
  }

  for (const ticker of bybitTickers) {
    const parsed = bybitApi.parseOptionSymbol(ticker.symbol)
    if (!parsed) continue
    const key = `${parsed.expiryDate}|${parsed.strikePrice}|${parsed.optionType.toLowerCase()}`
    const entry = ensure(key)
    const bid = parseFloat(ticker.bid1Price || 0)
    const ask = parseFloat(ticker.ask1Price || 0)
    entry.prices.bybit = { bid, ask }
    if (bid > entry.bestBid) { entry.bestBid = bid; entry.bestBidEx = 'bybit' }
    if (ask > 0 && (entry.bestAsk === 0 || ask < entry.bestAsk)) { entry.bestAsk = ask; entry.bestAskEx = 'bybit' }
    const bybitFwd = parseFloat(ticker.underlyingPrice || 0)
    if (bybitFwd > 0 && !forwardPrices[parsed.expiryDate]) forwardPrices[parsed.expiryDate] = bybitFwd
    entry.delta = parseFloat(ticker.delta || 0)
    entry.gamma = parseFloat(ticker.gamma || 0)
    entry.theta = parseFloat(ticker.theta || 0)
    entry.vega = parseFloat(ticker.vega || 0)
    const bMarkVol = parseFloat(ticker.impliedVolatility || 0)
    const bBidVol = parseFloat(ticker.bid1Iv || 0)
    const bAskVol = parseFloat(ticker.ask1Iv || 0)
    if (bMarkVol > 0) entry.markVol = bMarkVol
    if (bBidVol > 0) entry.bidVol = bBidVol
    if (bAskVol > 0) entry.askVol = bAskVol
  }

  for (const [instId, item] of Object.entries(okxFamilyCache)) {
    const parsed = parseOkxInstId(instId)
    if (!parsed) continue
    const key = `${parsed.expiryDate}|${parsed.strikePrice}|${parsed.optionType}`
    const entry = ensure(key)
    const ticker = okxFamilyTickers[instId] ?? {}
    const bid = parseFloat(ticker.bidPx || 0) * (okxSpot || 1)
    const ask = parseFloat(ticker.askPx || 0) * (okxSpot || 1)
    entry.prices.okx = { bid, ask }
    if (bid > entry.bestBid) { entry.bestBid = bid; entry.bestBidEx = 'okx' }
    if (ask > 0 && (entry.bestAsk === 0 || ask < entry.bestAsk)) { entry.bestAsk = ask; entry.bestAskEx = 'okx' }
    const okxFwd = parseFloat(item.fwdPx || 0)
    if (okxFwd > 0 && !forwardPrices[parsed.expiryDate]) forwardPrices[parsed.expiryDate] = okxFwd
    const delta = parseFloat(item.delta || 0)
    if (delta !== 0) {
      entry.delta = delta
      entry.gamma = okxSpot > 0 ? parseFloat(item.gamma || 0) / okxSpot : 0
      entry.theta = parseFloat(item.theta || 0) * okxSpot
      entry.vega = parseFloat(item.vega || 0) * okxSpot
    }
    const oMarkVol = parseFloat(item.markVol || 0)
    const oBidVol = parseFloat(item.bidVol || 0)
    const oAskVol = parseFloat(item.askVol || 0)
    if (oMarkVol > 0) entry.markVol = oMarkVol
    if (oBidVol > 0) entry.bidVol = oBidVol
    if (oAskVol > 0) entry.askVol = oAskVol
  }

  const deribitResp = buildDeribitResponse(baseCoin)
  if (deribitResp) {
    for (const [expiryDate, dateData] of Object.entries(deribitResp.data)) {
      if (dateData.forwardPrice > 0 && !forwardPrices[expiryDate]) {
        forwardPrices[expiryDate] = dateData.forwardPrice
      }
      for (const contract of [...dateData.calls, ...dateData.puts]) {
        const key = `${expiryDate}|${contract.strike}|${contract.optionType}`
        const entry = ensure(key)
        entry.prices.deribit = { bid: contract.bid, ask: contract.ask }
        if (contract.bid > entry.bestBid) { entry.bestBid = contract.bid; entry.bestBidEx = 'deribit' }
        if (contract.ask > 0 && (entry.bestAsk === 0 || contract.ask < entry.bestAsk)) { entry.bestAsk = contract.ask; entry.bestAskEx = 'deribit' }
        if (contract.delta !== 0) {
          entry.delta = contract.delta
          entry.gamma = contract.gamma
          entry.theta = contract.theta
          entry.vega = contract.vega
        }
        if (contract.markVol > 0) entry.markVol = contract.markVol
        if (contract.bidVol > 0) entry.bidVol = contract.bidVol
        if (contract.askVol > 0) entry.askVol = contract.askVol
      }
    }
  }

  if (deriveResp) {
    for (const [expiryDate, dateData] of Object.entries(deriveResp.data)) {
      for (const contract of [...dateData.calls, ...dateData.puts]) {
        const key = `${expiryDate}|${contract.strike}|${contract.optionType}`
        const entry = ensure(key)
        entry.prices.derive = { bid: contract.bid, ask: contract.ask }
        if (contract.bid > entry.bestBid) { entry.bestBid = contract.bid; entry.bestBidEx = 'derive' }
        if (contract.ask > 0 && (entry.bestAsk === 0 || contract.ask < entry.bestAsk)) { entry.bestAsk = contract.ask; entry.bestAskEx = 'derive' }
        if (contract.delta !== 0) {
          entry.delta = contract.delta
          entry.gamma = contract.gamma
          entry.theta = contract.theta
          entry.vega = contract.vega
        }
        if (contract.markVol > 0) entry.markVol = contract.markVol
        if (contract.bidVol > 0) entry.bidVol = contract.bidVol
        if (contract.askVol > 0) entry.askVol = contract.askVol
      }
    }
  }

  for (const [symbol, item] of Object.entries(binanceCache[baseCoin] ?? {})) {
    const parsed = parseBinanceSymbol(symbol)
    if (!parsed) continue
    const key = `${parsed.expiryDate}|${parsed.strikePrice}|${parsed.optionType}`
    const entry = ensure(key)
    const bid = parseFloat(item.bo ?? 0)
    const ask = parseFloat(item.ao ?? 0)
    entry.prices.binance = { bid, ask }
    if (bid > entry.bestBid) { entry.bestBid = bid; entry.bestBidEx = 'binance' }
    if (ask > 0 && (entry.bestAsk === 0 || ask < entry.bestAsk)) { entry.bestAsk = ask; entry.bestAskEx = 'binance' }
    const delta = parseFloat(item.d ?? 0)
    if (delta !== 0) {
      entry.delta = delta
      entry.gamma = parseFloat(item.g ?? 0)
      entry.theta = parseFloat(item.t ?? 0)
      entry.vega = parseFloat(item.v ?? 0)
    }
    const bMarkVol = parseFloat(item.vo ?? 0)
    const bBidVol = parseFloat(item.b ?? 0)
    const bAskVol = parseFloat(item.a ?? 0)
    if (bMarkVol > 0) entry.markVol = bMarkVol
    if (bBidVol > 0) entry.bidVol = bBidVol
    if (bAskVol > 0) entry.askVol = bAskVol
    const bnSpot = binanceSpotCache[baseCoin] ?? 0
    if (bnSpot > 0 && !forwardPrices[parsed.expiryDate]) forwardPrices[parsed.expiryDate] = bnSpot
  }

  const optionsByDate = {}
  for (const [key, entry] of Object.entries(merged)) {
    const [expiryDate, strikeStr, optionType] = key.split('|')
    const contract = { strike: parseFloat(strikeStr), optionType, ...entry }
    if (!optionsByDate[expiryDate]) {
      optionsByDate[expiryDate] = { calls: [], puts: [], forwardPrice: forwardPrices[expiryDate] ?? 0 }
    }
    if (optionType === 'call') optionsByDate[expiryDate].calls.push(contract)
    else optionsByDate[expiryDate].puts.push(contract)
  }

  const sortedDates = Object.keys(optionsByDate).sort()
  const expirationCounts = {}
  for (const date of sortedDates) {
    expirationCounts[date] = { calls: optionsByDate[date].calls.length, puts: optionsByDate[date].puts.length }
  }
  return { spotPrice, expirations: sortedDates, expirationCounts, data: optionsByDate }
}
